haystack/components/preprocessors/recursive_splitter.py

```diff
@@ -0,0 +1,296 @@
+            if curr_separator == "sentence":
+                sentence_with_spans = self.nltk_tokenizer.split_sentences(text)
+                splits = [sentence["sentence"] for sentence in sentence_with_spans]
+            else:
+                escaped_separator = re.escape(curr_separator)
+                escaped_separator = (
+                    f"({escaped_separator})"  # wrap the separator in a group to include it in the splits
+                )
+                splits = re.split(escaped_separator, text)
+
+                # merge every two consecutive splits, i.e.: the text and the separator after it
+                splits = [
+                    "".join([splits[i], splits[i + 1]]) if i < len(splits) - 1 else splits[i]
+                    for i in range(0, len(splits), 2)
+                ]
+
+                # remove last split if it is empty
+                splits = splits[:-1] if splits[-1] == "" else splits
+
+            if len(splits) == 1:  # go to next separator, if current separator not found in the text
+                continue
+
+            chunks = []
+            current_chunk: List[str] = []
+            current_length = 0
+
+            # check splits, if any is too long, recursively chunk it, otherwise add to current chunk
+            for split in splits:
+                split_text = split
+                # if adding this split exceeds chunk_size, process current_chunk
+                if current_length + self._chunk_length(split_text) > self.split_length:
+                    if current_chunk:  # keep the good splits
+                        chunks.append("".join(current_chunk))
+                        current_chunk = []
+                        current_length = 0
+
+                    # recursively handle splits that are too large
+                    if self._chunk_length(split_text) > self.split_length:
+                        if curr_separator == self.separators[-1]:
+                            # tried the last separator, can't split further, break the loop and fall back to
+                            # character-level chunking
+                            break
+                        chunks.extend(self._chunk_text(split_text))
+                    else:
+                        chunks.append(split_text)
+                else:
+                    current_chunk.append(split_text)
+                    current_length += self._chunk_length(split_text)
+
+            if current_chunk:
+                chunks.append("".join(current_chunk))
+
+            if self.split_overlap > 0:
+                chunks = self._apply_overlap(chunks)
+
+            if chunks:
+                return chunks
+
+        # if no separator worked, fall back to character- or word-level chunking
+        return [
+            text[i : i + self.split_length]
+            for i in range(0, self._chunk_length(text), self.split_length - self.split_overlap)
+        ]
```

Comment:
I like this idea! And I like that it's meant to default to character or word level depending on what the user chooses for the `split_units`. However, the way it is currently set up it only does chunking at the character level. And in this case if `split_units` is `word` then this would cause the text to be truncated since the `self._chunk_length(text)` does correctly measure the number of words when using `split_units=word`

Reply:
Thanks, I was aware of this and already fixed a few days ago - but I hadn't committed yet. Let me propose the following to ease both your work and mine, I suggest you start the review only after I issue a re-requested here - it can save us both time.

Reply:
Ahh okay, and sounds good! I'll wait for you! :)